cassiopeia/solver/HybridSolver.py

```diff
@@ -144,43 +165,58 @@ def solve(self):
 
             subproblem_tree = nx.relabel_nodes(subproblem_tree, mapping)
 
-            self.tree = nx.compose(self.tree, subproblem_tree)
+            self.__tree = nx.compose(self.__tree, subproblem_tree)
 
         # Collapse 0-mutation edges and append duplicate samples
-        self.tree = solver_utilities.collapse_tree(
-            self.tree, True, self.unique_character_matrix, self.missing_char
+        self.__tree = solver_utilities.collapse_tree(
+            self.__tree,
+            True,
+            unique_character_matrix,
+            cassiopeia_tree.missing_state_indicator,
         )
-        self.tree = self.append_sample_names(self.tree)
+        self.__tree = self.append_sample_names(self.__tree, character_matrix)
+
+        cassiopeia_tree.populate_tree(self.__tree)
 
     def apply_top_solver(
-        self, samples: List[str], root: Optional[int] = None
+        self,
+        character_matrix: pd.DataFrame,
+        samples: List[str],
+        weights: Optional[Dict[int, Dict[int, float]]] = None,
+        missing_state_indicator: int = -1,
+        root: Optional[int] = None,
     ) -> Tuple[int, List[Tuple[int, List[str]]]]:
         """Applies the top solver to samples.
 
-        A private helper method for applying the top solver to the samples
+        A helper method for applying the top solver to the samples
         until a criteria is hit. Subproblems and the root ID are returned.
 
         Args:
+            character_matrix: Character matrix
+            samples: Samples in the subclade of interest.
+            weights: Weights of character-state combinations, derived from
+                priors if these are available.
+            missing_state_indicator: Indicator for missing data
             root: Node ID of the root in the subtree containing the samples.
-            samples: Samples in the subclade.
 
         Returns:
             The ID of the node serving as the root of the tree containing the
                 samples, and a list of subproblems in the form
                 [subtree-root, subtree-samples].
         """
-        clades = list(self.top_solver.perform_split(samples))
 
-        root = (
-            len(self.tree.nodes)
-            - self.unique_character_matrix.shape[0]
-            + self.character_matrix.shape[0]
+        clades = list(
+            self.top_solver.perform_split(
+                character_matrix, samples, weights, missing_state_indicator
+            )
         )
 
-        self.tree.add_node(root)
+        root = len(self.__tree.nodes) + 1
+
+        self.__tree.add_node(root)
         if len(clades) == 1:
             for clade in clades[0]:
-                self.tree.add_edge(root, clade)
+                self.__tree.add_edge(root, clade)
```

Comment:
Just curious: is there a reason to use double underscore vs single underscore for class attributes? Or is there no difference?

Reply:
I think the difference here is between "protected" and "private" attributes/functions. (https://www.tutorialsteacher.com/python/public-private-protected-modifiers)

I think variables that begin with a single underscore can be accessed publicly but not changed, whereas double underscores indicates that they should not be accessed publicly.

Reply:
Ah, nice! Then double underscore is better. I didn't know about this!